src/rust/engine/src/externs/fs.rs

```diff
@@ -507,46 +517,104 @@ impl PyPathMetadata {
         Ok(self.0.path.clone())
     }
 
+    #[setter]
+    pub fn set_path(&mut self, path: PathBuf) -> PyResult<()> {
```

Comment:
I'm not super enthusiastic about setters that mutate an instance directly... I find reasoning about mutation more awkward than immutability, and there doesn't seem to be any existing uses of `#[setter]` in the codebase (but I'm not at my main machine so don't have the repo locally to confirm).

Some alternatives might be:

1. exposing a constructor, so Python can construct new instances directly (I think this would make the tests for `_path_metadata_to_bytes` simpler, since they wouldn't need to bounce through a rule runner request just to construct appropriate input for it)
2. using "with" methods that return a new instance, e.g. `with_path(&self, path: PathBuf) -> PyResult<PyPathMetadata>`

I think I prefer the first.

Given these are just designed for tests, I don't think it's worth blocking this PR on this.

Reply:
Good point. Immutable is to be preferred. Just pushed a change with a constructor and removed the setters.